### PR TITLE
Add `Android` support

### PR DESCRIPTION
**Pull request by yassernasc**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -23,7 +23,19 @@ target_sources(
...
 
-if(APPLE)
+if(ANDROID)
+  target_sources(
+    log
+    PRIVATE
+      src/android.c
+  )
+
+  target_link_libraries(
```
Comment (yassernasc) [inline CMakeLists.txt:33]: It was a solution for: https://stackoverflow.com/questions/4455941/undefined-reference-to-android-log-print
probably not the best approach

File: src/android.c [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+#include <android/log.h>
+
+#include "../include/log.h"
+
+typedef struct log_s log_t;
+
+struct log_s {
+  char *name;
+  int flags;
```
Comment (kasperisager) [inline src/android.c:13]:
```suggestion
```

File: src/android.c [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+static log_t *log_;
+
+int
+log_open (const char *name, int flags) {
+  if (log_ != NULL) return -1;
+
+  log_ = malloc(sizeof(log_t));
+  log_->name = strdup(name);
+  log_->flags = flags;
```
Comment (kasperisager) [inline src/android.c:24]:
```suggestion
```

File: src/android.c [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  if (err < 0) return err;
+
+  err = __android_log_vprint(ANDROID_LOG_ERROR, log_->name, "%s", args);
+
+  return err == 1 ? 0 : err;
+}
+
+int
+log_vfatal (const char *message, va_list args) {
+  if (log_ == NULL) goto done;
```
Comment (kasperisager) [inline src/android.c:124]:
```suggestion
  if (log_ == NULL) return -1;
```

File: src/android.c [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  *result = malloc(*size);
+
+  va_copy(args_copy, args);
+
+  vsnprintf(*result, *size, message, args);
+
+  va_end(args_copy);
+
+  return 0;
+}
```
Comment (kasperisager) [inline src/android.c:60]:
```suggestion
```

Android already support formatted printing so this function isn't required.

File: src/android.c [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+log_vdebug (const char *message, va_list args) {
+  if (log_ == NULL) return -1;
+
+  char *formatted;
+  size_t size;
+
+  int err = log_vformat(&formatted, &size, message, args);
+  if (err < 0) return err;
+
+  err = __android_log_vprint(ANDROID_LOG_DEBUG, log_->name, "%s", args);
```
Comment (kasperisager) [inline src/android.c:72]:
```suggestion
  int err = __android_log_vprint(ANDROID_LOG_DEBUG, log_->name, message, args);
```

File: src/android.c [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+log_vinfo (const char *message, va_list args) {
+  if (log_ == NULL) return -1;
+
+  char *formatted;
+  size_t size;
+
+  int err = log_vformat(&formatted, &size, message, args);
+  if (err < 0) return err;
+
+  err = __android_log_vprint(ANDROID_LOG_INFO, log_->name, "%s", args);
```
Comment (kasperisager) [inline src/android.c:87]:
```suggestion
  int err = __android_log_vprint(ANDROID_LOG_INFO, log_->name, message, args);
```

File: src/android.c [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+log_vwarn (const char *message, va_list args) {
+  if (log_ == NULL) return -1;
+
+  char *formatted;
+  size_t size;
+
+  int err = log_vformat(&formatted, &size, message, args);
+  if (err < 0) return err;
+
+  err = __android_log_vprint(ANDROID_LOG_WARN, log_->name, "%s", args);
```
Comment (kasperisager) [inline src/android.c:102]:
```suggestion
  int err = __android_log_vprint(ANDROID_LOG_WARN, log_->name, message, args);
```

File: src/android.c [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+log_verror (const char *message, va_list args) {
+  if (log_ == NULL) return -1;
+
+  char *formatted;
+  size_t size;
+
+  int err = log_vformat(&formatted, &size, message, args);
+  if (err < 0) return err;
+
+  err = __android_log_vprint(ANDROID_LOG_ERROR, log_->name, "%s", args);
```
Comment (kasperisager) [inline src/android.c:117]:
```suggestion
  int err = __android_log_vprint(ANDROID_LOG_ERROR, log_->name, message, args);
```

File: src/android.c [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+  int err = log_vformat(&formatted, &size, message, args);
+  if (err < 0) goto close;
+
+  __android_log_vprint(ANDROID_LOG_FATAL, log_->name, "%s", args);
+
+close:
+  log_close();
+
+done:
+  exit(1);
```
Comment (kasperisager) [inline src/android.c:138]:
```suggestion
  __android_log_vprint(ANDROID_LOG_FATAL, log_->name, message, args);

  exit(1);
```

File: CMakeLists.txt [hunk 11]
```diff
@@ -23,7 +23,19 @@ target_sources(
...
+    log
+    PRIVATE
+      -llog
+  )
```
Comment (kasperisager) [inline CMakeLists.txt:37]:
```suggestion
  find_library(log log)

  target_link_libraries(
    log
    PUBLIC
      ${log}
  )
```

**Review by kasperisager — CHANGES_REQUESTED** [review]

**Comment by yassernasc** [comment]
All comments are fixed, I think the only thing missing is to add Android support at `.github/workflows/test.yml`

**Review by kasperisager — APPROVED** [review]

**Comment by kasperisager** [comment]
Woop! 🥳

**Pull request merged**